File: results/dnd-rest-benchmark/lifecycle-runs/20260730T072747Z_pi_kimi-k2p7-code_go-stdlib/compendium.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
)

type createMonsterRequest struct {
	Slug       string   `json:"slug"`
	Name       string   `json:"name"`
	CR         string   `json:"cr"`
	ArmorClass int      `json:"armor_class"`
	HitPoints  int      `json:"hit_points"`
	Tags       []string `json:"tags"`
}

type monsterCreateResponse struct {
	Slug       string `json:"slug"`
	Name       string `json:"name"`
	CR         string `json:"cr"`
	ArmorClass int    `json:"armor_class"`
	HitPoints  int    `json:"hit_points"`
}

type monsterResponse struct {
	Slug       string   `json:"slug"`
	Name       string   `json:"name"`
	CR         string   `json:"cr"`
	ArmorClass int      `json:"armor_class"`
	HitPoints  int      `json:"hit_points"`
	Tags       []string `json:"tags"`
}

// dbMonster is the raw row shape returned by sqlite3 for compendium_monsters.
// Tags are stored as a JSON string.
type dbMonster struct {
	Slug       string `json:"slug"`
	Name       string `json:"name"`
	CR         string `json:"cr"`
	ArmorClass int    `json:"armor_class"`
	HitPoints  int    `json:"hit_points"`
	Tags       string `json:"tags"`
}

// itemRecord is both the request and response shape for compendium items.
type itemRecord struct {
	Slug   string `json:"slug"`
	Name   string `json:"name"`
	Type   string `json:"type"`
	Rarity string `json:"rarity"`
	CostGP int    `json:"cost_gp"`
}

// queryMonsterExists returns true when a monster with the given slug exists.
func queryMonsterExists(slug string) (bool, error) {
	return queryExists(fmt.Sprintf("SELECT 1 FROM compendium_monsters WHERE slug=%s LIMIT 1;", sq(slug)))
}

// queryMonster loads a single monster by slug, parsing its stored tags JSON.
// It returns nil when the monster is not found.
func queryMonster(slug string) (*monsterResponse, error) {
	var rows []dbMonster
	if err := queryRows(fmt.Sprintf("SELECT slug, name, cr, armor_class, hit_points, tags FROM compendium_monsters WHERE slug=%s LIMIT 1;", sq(slug)), &rows); err != nil {
		return nil, err
	}
	if len(rows) == 0 {
		return nil, nil
	}
	r := rows[0]
	tags := []string{}
	if r.Tags != "" && r.Tags != "null" {
		if err := json.Unmarshal([]byte(r.Tags), &tags); err != nil {
			return nil, err
		}
	}
	return &monsterResponse{
		Slug:       r.Slug,
		Name:       r.Name,
		CR:         r.CR,
		ArmorClass: r.ArmorClass,
		HitPoints:  r.HitPoints,
		Tags:       tags,
	}, nil
}

// queryItemExists returns true when an item with the given slug exists.
func queryItemExists(slug string) (bool, error) {
	return queryExists(fmt.Sprintf("SELECT 1 FROM compendium_items WHERE slug=%s LIMIT 1;", sq(slug)))
}

// createMonsterHandler stores a new monster entry in the compendium. The slug
// must be unique, and the response omits the tags list to match the original
// contract.
func createMonsterHandler(w http.ResponseWriter, r *http.Request) {
	var req createMonsterRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	if req.Slug == "" || req.Name == "" || req.CR == "" || req.ArmorClass <= 0 || req.HitPoints <= 0 {
		writeError(w, http.StatusBadRequest, "invalid monster")
		return
	}
	if req.Tags == nil {
		req.Tags = []string{}
	}
	tagsJSON, err := json.Marshal(req.Tags)
	if err != nil {
		log.Printf("monster tags marshal error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}

	dbMu.Lock()
	defer dbMu.Unlock()

	exists, err := queryMonsterExists(req.Slug)
	if err != nil {
		log.Printf("monster exists query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if exists {
		writeError(w, http.StatusConflict, "monster already exists")
		return
	}

	insertSQL := fmt.Sprintf("INSERT INTO compendium_monsters (slug, name, cr, armor_class, hit_points, tags) VALUES (%s, %s, %s, %d, %d, %s);",
		sq(req.Slug), sq(req.Name), sq(req.CR), req.ArmorClass, req.HitPoints, sq(string(tagsJSON)))
	if err := dbExec(insertSQL); err != nil {
		log.Printf("monster insert error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}

	writeJSON(w, http.StatusCreated, monsterCreateResponse{
		Slug:       req.Slug,
		Name:       req.Name,
		CR:         req.CR,
		ArmorClass: req.ArmorClass,
		HitPoints:  req.HitPoints,
	})
}

// getMonsterHandler reads a single monster from the compendium.
func getMonsterHandler(w http.ResponseWriter, r *http.Request) {
	slug := r.PathValue("slug")
	if slug == "" {
		writeError(w, http.StatusBadRequest, "missing slug")
		return
	}

	dbMu.Lock()
	defer dbMu.Unlock()

	monster, err := queryMonster(slug)
	if err != nil {
		log.Printf("monster get query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if monster == nil {
		writeError(w, http.StatusNotFound, "monster not found")
		return
	}

	writeJSON(w, http.StatusOK, monster)
}

// createItemHandler stores a new item entry in the compendium. The slug must
// be unique and all fields must be positive/non-empty.
func createItemHandler(w http.ResponseWriter, r *http.Request) {
	var req itemRecord
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	if req.Slug == "" || req.Name == "" || req.Type == "" || req.Rarity == "" || req.CostGP <= 0 {
		writeError(w, http.StatusBadRequest, "invalid item")
		return
	}

	dbMu.Lock()
	defer dbMu.Unlock()

	exists, err := queryItemExists(req.Slug)
	if err != nil {
		log.Printf("item exists query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if exists {
		writeError(w, http.StatusConflict, "item already exists")
		return
	}

	insertSQL := fmt.Sprintf("INSERT INTO compendium_items (slug, name, type, rarity, cost_gp) VALUES (%s, %s, %s, %s, %d);",
		sq(req.Slug), sq(req.Name), sq(req.Type), sq(req.Rarity), req.CostGP)
	if err := dbExec(insertSQL); err != nil {
		log.Printf("item insert error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}

	writeJSON(w, http.StatusCreated, itemRecord{
		Slug:   req.Slug,
		Name:   req.Name,
		Type:   req.Type,
		Rarity: req.Rarity,
		CostGP: req.CostGP,
	})
}

// getItemHandler reads a single item from the compendium.
func getItemHandler(w http.ResponseWriter, r *http.Request) {
	slug := r.PathValue("slug")
	if slug == "" {
		writeError(w, http.StatusBadRequest, "missing slug")
		return
	}

	dbMu.Lock()
	defer dbMu.Unlock()

	var items []itemRecord
	if err := queryRows(fmt.Sprintf("SELECT slug, name, type, rarity, cost_gp FROM compendium_items WHERE slug=%s LIMIT 1;", sq(slug)), &items); err != nil {
		log.Printf("item get query error: %v", err)
		writeError(w, http.StatusInternalServerError, "internal error")
		return
	}
	if len(items) == 0 {
		writeError(w, http.StatusNotFound, "item not found")
		return
	}

	writeJSON(w, http.StatusOK, items[0])
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260730T150315Z_pi_kimi-k2p7-code_php-stdlib/src/Compendium.php
<?php

declare(strict_types=1);

/**
 * Monster and item compendium entries.
 */

function createMonster(array $input): array
{
    $required = ['slug', 'name', 'cr', 'armor_class', 'hit_points'];
    foreach ($required as $key) {
        if (!array_key_exists($key, $input)) {
            sendError(400, 'missing fields');
        }
    }

    $slug = $input['slug'];
    $name = $input['name'];
    $cr = $input['cr'];
    if (!is_string($slug) || $slug === '' || !is_string($name) || $name === '' || !is_string($cr) || $cr === '') {
        sendError(400, 'invalid fields');
    }

    $armorClass = filter_var($input['armor_class'], FILTER_VALIDATE_INT);
    $hitPoints = filter_var($input['hit_points'], FILTER_VALIDATE_INT);
    if ($armorClass === false || $hitPoints === false) {
        sendError(400, 'invalid fields');
    }

    $tags = [];
    if (array_key_exists('tags', $input)) {
        if (!is_array($input['tags'])) {
            sendError(400, 'invalid tags');
        }
        foreach ($input['tags'] as $tag) {
            if (!is_string($tag)) {
                sendError(400, 'invalid tags');
            }
            $tags[] = $tag;
        }
    }

    if (getMonsterBySlug($slug) !== null) {
        sendError(409, 'slug already exists');
    }

    $stmt = db()->prepare('INSERT INTO monsters (slug, name, cr, armor_class, hit_points, tags_json) VALUES (?, ?, ?, ?, ?, ?)');
    $stmt->execute([$slug, $name, $cr, $armorClass, $hitPoints, json_encode($tags, JSON_UNESCAPED_SLASHES)]);

    return [
        'slug' => $slug,
        'name' => $name,
        'cr' => $cr,
        'armor_class' => $armorClass,
        'hit_points' => $hitPoints,
    ];
}

function getMonsterBySlug(string $slug): ?array
{
    $stmt = db()->prepare('SELECT slug, name, cr, armor_class, hit_points, tags_json FROM monsters WHERE slug = ?');
    $stmt->execute([$slug]);
    $row = $stmt->fetch(PDO::FETCH_ASSOC);
    if ($row === false) {
        return null;
    }
    return [
        'slug' => $row['slug'],
        'name' => $row['name'],
        'cr' => $row['cr'],
        'armor_class' => (int) $row['armor_class'],
        'hit_points' => (int) $row['hit_points'],
        'tags' => json_decode($row['tags_json'], true),
    ];
}

function createItem(array $input): array
{
    $required = ['slug', 'name', 'type', 'rarity', 'cost_gp'];
    foreach ($required as $key) {
        if (!array_key_exists($key, $input)) {
            sendError(400, 'missing fields');
        }
    }

    $slug = $input['slug'];
    $name = $input['name'];
    $type = $input['type'];
    $rarity = $input['rarity'];
    if (!is_string($slug) || $slug === '' || !is_string($name) || $name === '' || !is_string($type) || $type === '' || !is_string($rarity) || $rarity === '') {
        sendError(400, 'invalid fields');
    }

    $costGp = filter_var($input['cost_gp'], FILTER_VALIDATE_INT);
    if ($costGp === false) {
        sendError(400, 'invalid fields');
    }

    if (getItemBySlug($slug) !== null) {
        sendError(409, 'slug already exists');
    }

    $stmt = db()->prepare('INSERT INTO items (slug, name, type, rarity, cost_gp) VALUES (?, ?, ?, ?, ?)');
    $stmt->execute([$slug, $name, $type, $rarity, $costGp]);

    return [
        'slug' => $slug,
        'name' => $name,
        'type' => $type,
        'rarity' => $rarity,
        'cost_gp' => $costGp,
    ];
}

function getItemBySlug(string $slug): ?array
{
    $stmt = db()->prepare('SELECT slug, name, type, rarity, cost_gp FROM items WHERE slug = ?');
    $stmt->execute([$slug]);
    $row = $stmt->fetch(PDO::FETCH_ASSOC);
    if ($row === false) {
        return null;
    }
    return [
        'slug' => $row['slug'],
        'name' => $row['name'],
        'type' => $row['type'],
        'rarity' => $row['rarity'],
        'cost_gp' => (int) $row['cost_gp'],
    ];
}
